### DISCUSSION TITLE
Clipboard doesn't work on remote in nvim (wezterm connect doesn't set $DISPLAY?)

### DISCUSSION THREAD

**Discussion by kescobo** [wezterm/wezterm#5231] [Q&A] [answered]
### What Operating System(s) are you running on?

Linux X11

### Which Wayland compositor or X11 Window manager(s) are you using?

KDE Plasma 6

### WezTerm version

20240203.110809.5046fc22

### Ask your question!

I'm trying to use wezterm multiplexer over SSH to the lab server, but I'm running into problems with yanking to the system clipboard in neovim. It might be because, when using `wezterm connect`, it seems like the `$DISPLAY` isn't set, though setting it manually doesn't seem to fix the problem.

## Steps to reproduce

- use `wezterm connect SSHMUX:myserver`
- open nvim
- try to yank to system clipboard

Get: `clipboard: No provider. Try ":checkhealth" or ":h clipboard"

`echo $DISPLAY` gives no output

## Fixes

If I connect via `ssh myserver`, this doesn't happen. When I `echo $DISPLAY` in this case, I get `localhost:10.0`, though I've tried manually setting this variable when using `wezterm connect`, and I still can't use the system clipboard in nvim

**Comment by kescobo** [comment]
Some other notes:

Here is the clipboard-relevant output from `:checkhealth` in nvim:

```
13 Clipboard (optional)
   12 - WARNING No clipboard tool found. Clipboard registers ("+ and "*) will not work.
   11   - ADVICE:
   10     - :help clipboard
```

I was able to copy that using wezterm's copy-mode (ctrl+shift+x, hightlight, y), though not with normal terminal `ctrl+shift+c`

**Comment by wez** [comment, marked as answer]
wezterm doesn't support X forwarding over its ssh connections.
For remote clipboard that works regardless, I recommend https://github.com/ojroques/vim-oscyank

**Reply by kescobo** [reply]
That package says it's obsolete as of Neovim 0.10. Current release is still 0.9.x, so I downloaded and installed the nightly neovim, and added
```
      vim.g.clipboard = {
        name = 'OSC 52',
        copy = {
          ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
          ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
        },
        paste = {
          ['+'] = require('vim.ui.clipboard.osc52').paste('+'),
          ['*'] = require('vim.ui.clipboard.osc52').paste('*'),
        },
      }
```

to the config as suggested... somewhere. Can't find the source, but it works beautifully :+1: 

Thanks for the prompt response!

**Reply by byddha** [reply]
With this yanking will work, but not pasting, since wezterm only allows changing the clipboard, but not reading it.

"Requests to query the clipboard are ignored. Allows setting or clearing the clipboard" https://wezfurlong.org/wezterm/escape-sequences.html#operating-system-command-sequences.

So if X forwarding is not supported and OSC 52 queries are not supported, i don't know if there is any workaround around this unfortunately

L.E: This is the kinda annoying workaround I found. Inspired from https://github.com/neovim/neovim/issues/28611, i've added this to my nvim config:

```lua
if os.getenv "SSH_CLIENT" ~= nil or os.getenv "SSH_TTY" ~= nil then
    local function my_paste(_)
        return function(_)
            local content = vim.fn.getreg '"'
            return vim.split(content, "\n")
        end
    end

    vim.g.clipboard = {
        name = "OSC 52",
        copy = {
            ["+"] = require("vim.ui.clipboard.osc52").copy "+",
            ["*"] = require("vim.ui.clipboard.osc52").copy "*",
        },
        paste = {
            ["+"] = my_paste "+",
            ["*"] = my_paste "*",
        },
    }
end
```

And for whenever I need to paste something from the host to the ssh client, I use wezterm PasteFrom (you could just use ctrl+shift+v)

```lua
{
    key = "p",
    mods = "ALT",
    action = act.PasteFrom("Clipboard"),
},

```

**Reply by joncrangle** [reply]
This is great, thanks for sharing. 

EDIT:
I've added a variation of this in my config to also detect Wezterm remote SSH sessions using the WEZTERM_EXECUTABLE environment variable. 
 
```lua
-- Sync clipboard between OS and Neovim.
-- Function to set OSC 52 clipboard
local function set_osc52_clipboard()
  local function my_paste()
    local content = vim.fn.getreg '"'
    return vim.split(content, '\n')
  end

  vim.g.clipboard = {
    name = 'OSC 52',
    copy = {
      ['+'] = require('vim.ui.clipboard.osc52').copy '+',
      ['*'] = require('vim.ui.clipboard.osc52').copy '*',
    },
    paste = {
      ['+'] = my_paste,
      ['*'] = my_paste,
    },
  }
end

-- Check if the current session is a remote WezTerm session based on the WezTerm executable
local function check_wezterm_remote_clipboard(callback)
  local wezterm_executable = vim.uv.os_getenv 'WEZTERM_EXECUTABLE'

  if wezterm_executable and wezterm_executable:find('wezterm-mux-server', 1, true) then
    callback(true) -- Remote WezTerm session found
  else
    callback(false) -- No remote WezTerm session
  end
end

-- Schedule the setting after `UiEnter` because it can increase startup-time.
vim.schedule(function()
  vim.opt.clipboard:append 'unnamedplus'

  -- Standard SSH session handling
    if vim.uv.os_getenv 'SSH_CLIENT' ~= nil or vim.uv.os_getenv 'SSH_TTY' ~= nil then
    set_osc52_clipboard()
  else
    check_wezterm_remote_clipboard(function(is_remote_wezterm)
      if is_remote_wezterm then
        set_osc52_clipboard()
      end
    end)
  end
end)
```
